// wiremock-client/dist/wiremock-client.esm.js
import axios from 'axios';

function createCommonjsModule(fn, module) {
	return module = { exports: {} }, fn(module, module.exports), module.exports;
}

var mappingBuilder = class MappingBuilder {
  constructor(method, url) {
    this.method = method;
    this.url = url;
    this.headers = {};
    this.cookies = {};
    this.queryParameters = {};
    this.bodyPatterns = [];
  }

  atPriority(priority) {
    this.priority = priority;
    return this;
  }

  build() {
    const mapping = {
      request: {
        method: this.method
      }
    };

    if (this.priority) {
      mapping.priority = this.priority;
    } // request


    Object.assign(mapping.request, this.url);

    if (Object.keys(this.headers).length > 0) {
      mapping.request.headers = this.headers;
    }

    if (Object.keys(this.cookies).length > 0) {
      mapping.request.cookies = this.cookies;
    }

    if (Object.keys(this.queryParameters).length > 0) {
      mapping.request.queryParameters = this.queryParameters;
    }

    if (this.bodyPatterns.length > 0) {
      mapping.request.bodyPatterns = this.bodyPatterns;
    } // response


    if (this.response) {
      mapping.response = this.response.build();
    }

    return mapping;
  }

  withHeader(key, value) {
    this.headers[key] = value;
    return this;
  }

  withCookie(key, value) {
    this.cookies[key] = value;
    return this;
  }

  withQueryParam(key, value) {
    this.queryParameters[key] = value;
    return this;
  }

  withRequestBody(bodyPattern) {
    this.bodyPatterns.push(bodyPattern);
    return this;
  }

  willReturn(responseBuilder) {
    this.response = responseBuilder;
    return this;
  }

};

function notNumeric(val) {
  return typeof val !== 'number' || Number.isNaN(val);
}

var countMatchers = {
  lessThan(expected) {
    return actual => {
      if (notNumeric(actual)) {
        return false;
      }

      return expected > actual;
    };
  },

  lessThanOrExactly(expected) {
    return actual => {
      if (notNumeric(actual)) {
        return false;
      }

      return expected >= actual;
    };
  },

  exactly(expected) {
    return actual => {
      if (notNumeric(actual)) {
        return false;
      }

      return expected === actual;
    };
  },

  moreThanOrExactly(expected) {
    return actual => {
      if (notNumeric(actual)) {
        return false;
      }

      return expected <= actual;
    };
  },

  moreThan(expected) {
    return actual => {
      if (notNumeric(actual)) {
        return false;
      }

      return expected < actual;
    };
  }

};

var urlMatchers = {
  urlEqualTo(url) {
    return {
      url
    };
  },

  urlMatching(urlPattern) {
    return {
      urlPattern
    };
  },

  urlPathEqualTo(urlPath) {
    return {
      urlPath
    };
  },

  urlPathMatching(urlPathPattern) {
    return {
      urlPathPattern
    };
  }

};

var matchers = {
  absent(absent) {
    return {
      absent
    };
  },

  binaryEqualTo(binaryEqualTo) {
    return {
      binaryEqualTo
    };
  },

  containing(contains) {
    return {
      contains
    };
  },

  equalTo(equalTo) {
    return {
      equalTo
    };
  },

  equalToIgnoreCase(equalTo) {
    return {
      equalTo,
      caseInsensitive: true
    };
  },

  equalToJson(equalToJson, ignoreArrayOrder, ignoreExtraElements) {
    let matcher = {
      equalToJson
    };

    if (ignoreArrayOrder) {
      matcher.ignoreArrayOrder = true;
    }

    if (ignoreExtraElements) {
      matcher.ignoreExtraElements = true;
    }

    return matcher;
  },

  equalToXml(equalToXml) {
    return {
      equalToXml
    };
  },

  matchesJsonPath(matchesJsonPath) {
    return {
      matchesJsonPath
    };
  },

  matchingXPath(matchingXPath) {
    return {
      matchingXPath
    };
  },

  matches(matches) {
    return {
      matches
    };
  },

  notMatching(doesNotMatch) {
    return {
      doesNotMatch
    };
  }

};

var matchers$1 = createCommonjsModule(function (module) {
Object.assign(module.exports, countMatchers);
Object.assign(module.exports, urlMatchers);
Object.assign(module.exports, matchers);
});

const {
  absent,
  urlEqualTo
} = matchers$1;

var requestPatternBuilder = class RequestPatternBuilder {
  constructor(method, url) {
    this.method = method;
    this.url = url;
    this.headers = {};
    this.cookies = {};
    this.queryParameters = {};
    this.bodyPatterns = [];
  }

  build() {
    const requestPattern = {
      method: this.method
    };
    Object.assign(requestPattern, this.url);

    if (Object.keys(this.headers).length > 0) {
      requestPattern.headers = this.headers;
    }

    if (Object.keys(this.cookies).length > 0) {
      requestPattern.cookies = this.cookies;
    }

    if (Object.keys(this.queryParameters).length > 0) {
      requestPattern.queryParameters = this.queryParameters;
    }

    if (this.bodyPatterns.length > 0) {
      requestPattern.bodyPatterns = this.bodyPatterns;
    }

    if (this.basicAuthCredentials) {
      requestPattern.basicAuthCredentials = this.basicAuthCredentials;
    }

    return requestPattern;
  }

  withUrl(url) {
    this.url = urlEqualTo(url);
    return this;
  }

  withHeader(key, value) {
    this.headers[key] = value;
    return this;
  }

  withoutHeader(key) {
    this.headers[key] = absent();
    return this;
  }

  withCookie(key, value) {
    this.cookies[key] = value;
    return this;
  }

  withQueryParam(key, value) {
    this.queryParameters[key] = value;
    return this;
  }

  withBasicAuth(username, password) {
    this.basicAuthCredentials = {
      username,
      password
    };
    return this;
  }

  withRequestBody(bodyPattern) {
    this.bodyPatterns.push(bodyPattern);
    return this;
  }

};

var responseBuilder = class ResponseBuilder {
  constructor() {
    this.headers = {};
  }

  build() {
    const response = {
      status: this.status
    };

    if (this.body) {
      response.body = this.body;
    }

    if (Object.keys(this.headers).length > 0) {
      response.headers = this.headers;
    }

    if (this.fault) {
      response.fault = this.fault;
    } // delay


    if (this.chunkedDribbleDelay) {
      response.chunkedDribbleDelay = this.chunkedDribbleDelay;
    }

    if (this.fixedDelayMilliseconds) {
      response.fixedDelayMilliseconds = this.fixedDelayMilliseconds;
    }

    if (this.delayDistribution) {
      response.delayDistribution = this.delayDistribution;
    }

    return response;
  }

  withFault(fault) {
    this.fault = fault;
    return this;
  }

  withBody(body) {
    this.body = typeof body === 'string' ? body : JSON.stringify(body);
    return this;
  }

  withHeader(key, value) {
    this.headers[key] = value;
    return this;
  }
  /*
   * Delay
   */


  withRandomChunkedDribbleDelay(numberOfChunks, totalDuration) {
    this.chunkedDribbleDelay = {
      numberOfChunks,
      totalDuration
    };
    return this;
  }

  withFixedDelay(fixedDelayMilliseconds) {
    this.fixedDelayMilliseconds = fixedDelayMilliseconds;
    return this;
  }

  withLogNormalRandomDelay(median, sigma) {
    this.delayDistribution = {
      type: 'lognormal',
      median,
      sigma
    };
    return this;
  }

  withUniformRandomDelay(lower, upper) {
    this.delayDistribution = {
      type: 'uniform',
      lower,
      upper
    };
    return this;
  }

  withStatus(status) {
    this.status = status;
    return this;
  }

};

const RequestMethods = {
  GET: 'GET',
  POST: 'POST',
  PUT: 'PUT',
  DELETE: 'DELETE',
  PATCH: 'PATCH',
  OPTIONS: 'OPTIONS',
  HEAD: 'HEAD',
  TRACE: 'TRACE',
  ANY: 'ANY'
};
var builders = {
  get: function (urlPattern) {
    return new mappingBuilder(RequestMethods.GET, urlPattern);
  },

  getRequestedFor(urlPattern) {
    return new requestPatternBuilder(RequestMethods.GET, urlPattern);
  },

  post: function (urlPattern) {
    return new mappingBuilder(RequestMethods.POST, urlPattern);
  },

  postRequestedFor(urlPattern) {
    return new requestPatternBuilder(RequestMethods.POST, urlPattern);
  },

  put: function (urlPattern) {
    return new mappingBuilder(RequestMethods.PUT, urlPattern);
  },

  putRequestedFor(urlPattern) {
    return new requestPatternBuilder(RequestMethods.PUT, urlPattern);
  },

  delete: function (urlPattern) {
    return new mappingBuilder(RequestMethods.DELETE, urlPattern);
  },

  deleteRequestedFor(urlPattern) {
    return new requestPatternBuilder(RequestMethods.DELETE, urlPattern);
  },

  patch: function (urlPattern) {
    return new mappingBuilder(RequestMethods.PATCH, urlPattern);
  },

  patchRequestedFor(urlPattern) {
    return new requestPatternBuilder(RequestMethods.PATCH, urlPattern);
  },

  options: function (urlPattern) {
    return new mappingBuilder(RequestMethods.OPTIONS, urlPattern);
  },

  optionsRequestedFor(urlPattern) {
    return new requestPatternBuilder(RequestMethods.OPTIONS, urlPattern);
  },

  head: function (urlPattern) {
    return new mappingBuilder(RequestMethods.HEAD, urlPattern);
  },

  headRequestedFor(urlPattern) {
    return new requestPatternBuilder(RequestMethods.HEAD, urlPattern);
  },

  trace: function (urlPattern) {
    return new mappingBuilder(RequestMethods.TRACE, urlPattern);
  },

  traceRequestedFor(urlPattern) {
    return new requestPatternBuilder(RequestMethods.TRACE, urlPattern);
  },

  any: function (urlPattern) {
    return new mappingBuilder(RequestMethods.ANY, urlPattern);
  },

  anyRequestedFor(urlPattern) {
    return new requestPatternBuilder(RequestMethods.ANY, urlPattern);
  },

  aResponse: function () {
    return new responseBuilder();
  }
};

var faults = {
  EMPTY_RESPONSE: 'EMPTY_RESPONSE',
  MALFORMED_RESPONSE_CHUNK: 'MALFORMED_RESPONSE_CHUNK',
  RANDOM_DATA_THEN_CLOSE: 'RANDOM_DATA_THEN_CLOSE',
  CONNECTION_RESET_BY_PEER: 'CONNECTION_RESET_BY_PEER',
  TIME_WAIT: 'TIME_WAIT'
};

var builders$1 = createCommonjsModule(function (module) {
Object.assign(module.exports, builders);
Object.assign(module.exports, faults);
Object.assign(module.exports, mappingBuilder);
Object.assign(module.exports, responseBuilder);
});

function handleResponse(promise) {
  return promise.then(res => res.data);
}

var httpAdminInterface = class HttpAdminInterface {
  constructor(protocol, host, port) {
    this.uri = `${protocol.toLowerCase()}://${host}:${port}/__admin`;
  }

  healthcheck() {
    return this.getRecordingStatus().then(() => true).catch(() => false);
  }

  resetAll() {
    return handleResponse(axios({
      method: 'post',
      url: `${this.uri}/reset`
    }));
  }
  /*
   * MAPPINGS
   */


  createMapping(mapping) {
    return handleResponse(axios({
      method: 'post',
      url: `${this.uri}/mappings`,
      data: mapping
    }));
  }

  getStubMapping(uuid) {
    return handleResponse(axios({
      method: 'get',
      url: `${this.uri}/mappings/${uuid}`
    }));
  }

  listAllStubMappings() {
    return handleResponse(axios({
      method: 'get',
      url: `${this.uri}/mappings`
    }));
  }

  resetMappings() {
    return handleResponse(axios({
      method: 'post',
      url: `${this.uri}/mappings/reset`
    }));
  }
  /*
   * REQUESTS
   */


  countRequestsMatching(requestPattern) {
    return handleResponse(axios({
      method: 'post',
      url: `${this.uri}/requests/count`,
      data: requestPattern
    }));
  }

  findRequestsMatching(requestPattern) {
    return handleResponse(axios({
      method: 'post',
      url: `${this.uri}/requests/find`,
      data: requestPattern
    }));
  }

  listAllRequests() {
    return handleResponse(axios({
      method: 'get',
      url: `${this.uri}/requests`
    }));
  }

  resetRequests() {
    return handleResponse(axios({
      method: 'post',
      url: `${this.uri}/requests/reset`
    }));
  }
  /*
   * RECORDINGS
   */


  getRecordingStatus() {
    return handleResponse(axios({
      method: 'get',
      url: `${this.uri}/recordings/status`
    }));
  }

};

const {
  moreThanOrExactly
} = matchers$1;

const DEFAULT_PORT = 8080;
const DEFAULT_PROTOCOL = 'http';
const DEFAULT_HOST = 'localhost';

var connect = function connect(protocol, host, port) {
  const admin = new httpAdminInterface(protocol || DEFAULT_PROTOCOL, host || DEFAULT_HOST, port || DEFAULT_PORT);
  return {
    healthcheck() {
      return admin.healthcheck();
    },

    // POST /reset
    resetAll() {
      return admin.resetAll();
    },

    /*
     * MAPPINGS
     */
    // GET /mappings
    allStubMappings() {
      return admin.listAllStubMappings();
    },

    getStubMapping(uuid) {
      return admin.getStubMapping(uuid);
    },

    // POST /mappings
    register(mappingBuilder$$1) {
      if (mappingBuilder$$1 instanceof mappingBuilder) {
        return admin.createMapping(mappingBuilder$$1.build());
      }

      return admin.createMapping(mappingBuilder$$1);
    },

    // POST /mappings/reset
    resetMappings() {
      return admin.resetMappings();
    },

    /*
     * REQUESTS
     */
    allRequests() {
      return admin.listAllRequests();
    },

    find(requestBuilder) {
      if (requestBuilder instanceof requestPatternBuilder) {
        return admin.findRequestsMatching(requestBuilder.build());
      }

      return admin.findRequestsMatching(requestBuilder);
    },

    verifyThat(requestBuilder, countMatcher = moreThanOrExactly(1)) {
      if (requestBuilder instanceof requestPatternBuilder) {
        return admin.findRequestsMatching(requestBuilder.build()).then(res => {
          return Array.isArray(res.requests) && countMatcher(res.requests.length);
        });
      }

      return admin.findRequestsMatching(requestBuilder).then(res => {
        return Array.isArray(res.requests) && countMatcher(res.requests.length);
      });
    },

    // POST /requests/reset
    resetRequests() {
      return admin.resetRequests();
    },

    /*
     * RECORDINGS
     */
    recordingStatus() {
      return admin.getRecordingStatus();
    }

  };
};

var src = createCommonjsModule(function (module) {
// builders


module.exports = builders$1;
module.exports.Builders = builders$1; // matchers



Object.assign(module.exports, matchers$1);
module.exports.Matchers = matchers$1; // connect

module.exports.connect = connect;
});
var src_1 = src.Builders;
var src_2 = src.Matchers;
var src_3 = src.connect;

export default src;
export { src_1 as Builders, src_2 as Matchers, src_3 as connect };
